op-node/rollup/derive/span_channel_out.go

```diff
@@ -77,115 +114,119 @@ func (co *SpanChannelOut) AddBlock(rollupCfg *rollup.Config, block *types.Block)
-			co.reader.Reset()
-			return 0, fmt.Errorf("failed to copy compressed data to reader: %w", err)
-		}
+			co.activeRLP().Len(), co.inactiveRLP().Len(), MaxRLPBytesPerChannel, ErrTooManyRLPBytes)
 	}
 
-	// Reset compressor to rewrite the entire span batch
-	co.compress.Reset()
-	// Avoid using io.Copy here, because we need all or nothing
-	written, err := co.compress.Write(buf.Bytes())
-	// Always flush (for BlindCompressor to check if it's full)
-	if err := co.compress.Flush(); err != nil {
-		return 0, fmt.Errorf("failed to flush compressor: %w", err)
+	// if the compressed data *plus* the new rlp data is under the target size, return early
+	// this optimizes out cases where the compressor will obviously come in under the target size
+	rlpGrowth := co.activeRLP().Len() - co.lastCompressedRLPSize
+	if uint64(co.compressed.Len()+rlpGrowth) < co.target {
+		return uint64(co.lastCompressedRLPSize), nil
 	}
-	if co.compress.FullErr() != nil {
-		err = co.compress.FullErr()
+
+	// we must compress the data to check if we've met or exceeded the target size
+	if err = co.compress(); err != nil {
+		return 0, err
+	}
+	co.lastCompressedRLPSize = co.activeRLP().Len()
+
+	// if the channel is now full, either return the compressed data, or the compressed previous data
+	if err := co.FullErr(); err != nil {
+		// if there is only one batch in the channel, it *must* be returned
 		if len(co.spanBatch.Batches) == 1 {
-			// Do not return ErrCompressorFull for the first block in the batch
-			// In this case, reader must be empty. then the contents of compressor will be copied to reader when the channel is closed.
-			err = nil
+			return uint64(co.compressed.Len()), nil
+		}
+
+		// if there is more than one batch in the channel, we revert the last batch
+		// by switching the RLP buffer and doing a fresh compression
+		co.swapRLP()
+		if err = co.compress(); err != nil {
+			return 0, err
 		}
-		// If there are more than one blocks in the channel, reader should have data that preserves previous compression result before adding this block.
-		// So, as a result, this block is not added to the channel and the channel will be closed.
-		return uint64(written), err
+		// co.FullErr should always return FullErr by this point
+		return uint64(co.compressed.Len()), co.FullErr()
 	}
 
-	// If compressor is not full yet, reader must be reset to avoid submitting invalid frames
-	co.reader.Reset()
-	return uint64(written), err
+	return uint64(co.compressed.Len()), nil
+}
+
+// compress compresses the active RLP buffer and checks if the compressed data is over the target size.
+// it resets all the compression buffers because Span Batches aren't meant to be compressed incrementally.
+func (co *SpanChannelOut) compress() error {
+	co.compressed.Reset()
+	co.compressor.Reset(co.compressed)
+	if _, err := co.compressor.Write(co.activeRLP().Bytes()); err != nil {
+		return err
+	}
+	co.compressor.Flush()
+	co.checkFull()
```

Comment:
Couldn't we even just `Close` the compressor here, instead of just flushing? We could then also skip adding the `CloseOverheadZlib` constant in `checkFull` and just compare the compressed size and the target directly.

Reply:
Ah nice, I hadn't thought about that! Thanks :)